core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/PerformanceCounterContainer.java

```diff
@@ -0,0 +1,223 @@
+
+    private long startCollectingIntervalInMillis = START_COLLECTING_INTERVAL_IN_MILLIS;
+    private long collectingIntervalInMillis = COLLECTING_INTERVAL_IN_MILLIS;
+
+    private TelemetryClient telemetryClient;
+
+    private ScheduledThreadPoolExecutor threads;
+
+    /**
+     * Registers a {@link com.microsoft.applicationinsights.internal.perfcounter.PerformanceCounter} that can collect data.
+     * @param performanceCounter The Performance Counter.
+     */
+    public void register(PerformanceCounter performanceCounter) {
+        Preconditions.checkArgument(!Strings.isNullOrEmpty(performanceCounter.getId()), "performanceCounter's id should be non null, non empty value");
+
+        initialize();
+        performanceCounters.put(performanceCounter.getId(), performanceCounter);
+    }
+
+    /**
+     * Un-registers a performance counter.
+     * @param performanceCounter The Performance Counter.
+     */
+    public void unregister(PerformanceCounter performanceCounter) {
+        unregister(performanceCounter.getId());
+    }
+
+    /**
+     * Un-registers a performance counter by its id.
+     * @param id The Performance Counter's id.
+     */
+    public void unregister(String id) {
+        Preconditions.checkArgument(!Strings.isNullOrEmpty(id), "id should be non null, non empty value");
+
+        performanceCounters.remove(id);
+    }
+
+    /**
+     * Sets the timeout to wait before the first reporting.
+     *
+     * The number must be a positive number
+     *
+     * Note that the method will be effective if called before the first call to the 'register' method.
+     * @param startCollectingIntervalInMillis Timeout to wait before the first collection of performance counters in milliseconds.
+     */
+    public void setStartCollectingIntervalInMillis(long startCollectingIntervalInMillis) {
+        if (startCollectingIntervalInMillis <= 0) {
+            InternalLogger.INSTANCE.error("'%d' is an illegal value is ignored. Must be a positive number", startCollectingIntervalInMillis);
+            return;
+        }
+
+        this.startCollectingIntervalInMillis = startCollectingIntervalInMillis;
+    }
+
+    /**
+     * Gets the timeout in milliseconds that the container will wait before the first collection of Performance Counters.
+     * @return The first timeout in milliseconds.
+     */
+    public long getStartCollectingIntervalInMillis() {
+        return startCollectingIntervalInMillis;
+    }
+
+    /**
+     * Sets the timeout to wait between collection of Performance Counters.
+     *
+     * The number must be a positive number
+     *
+     * Note that the method will be effective if called before the first call to the 'register' method.
+     * @param collectingIntervalInMillis The timeout to wait between collection of Performance Counters.
+     */
+    public void setCollectingIntervalInMillis(long collectingIntervalInMillis) {
+        if (collectingIntervalInMillis <= 0) {
```

Comment:
I think that we need to enforce a minimal value of at least few seconds, otherwise if a user define 0.0001 we can get like an infinite loop here.